### PR TITLE
Support backward_passes_per_step > 1 for TF Keras Graph Mode

### PR DESCRIPTION
**Pull request by aaron276h**
## Checklist before submitting

- [x] Did you read the [contributor guide](https://github.com/horovod/horovod/blob/master/CONTRIBUTING.md)?
- [ ] Did you update the docs?
- [x] Did you write any tests to validate this change?  
- [x] Did you update the [CHANGELOG](https://github.com/horovod/horovod/blob/master/CHANGELOG.md), if this change affects users?

## Description
This PR adds support for `backward_passes_per_step > 1` for TF Keras optimizers executing in graph(non-eager) mode. This is one of the features that we have built into [Determined AI's](https://determined.ai/) fork of Horovod that we would like to upstream. After this PR we also plan to add support for `backward_passes_per_step > 1` for eager execution as well as Tensorflow optimizers (LegacyOptimizers).

### CODE REVIEW

File: horovod/common/gradient_aggregation.py [hunk 1]
```diff
@@ -0,0 +1,266 @@
+import tensorflow as tf
```
Comment (tgaddair) [inline horovod/common/gradient_aggregation.py:1]: Since `tensorflow` is imported here, can we move this to `horovod/tensorflow`?

Comment (aaron276h) [reply horovod/common/gradient_aggregation.py]: Moved to `horovod/tensorflow/gradient_aggregation.py`.

**Review by tgaddair — COMMENTED** [review]
Looks great!  Still going over the logic in the aggregation helper to make sure I understand all the implications.  Just a few minor things for now.

**Comment by tgaddair** [comment]
cc @romerojosh in case you'd like to review as well.

File: horovod/tensorflow/__init__.py [hunk 2]
```diff
@@ -434,8 +435,9 @@ def variables(self, *args, **kwargs):
 
 def DistributedOptimizer(optimizer, name=None, use_locking=False, device_dense='',
                          device_sparse='', compression=Compression.none,
-                         sparse_as_dense=False, backward_passes_per_step=1,
-                         op=Average, gradient_predivide_factor=1.0):
+                         backward_passes_per_step=1, sparse_as_dense=False,
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py]: Thanks for catching this, didn't mean to change the order.

File: horovod/tensorflow/__init__.py [hunk 3]
```diff
@@ -477,6 +479,10 @@ def DistributedOptimizer(optimizer, name=None, use_locking=False, device_dense='
         before and after the sum. Gradients are scaled by
         1.0 / gradient_predivide_factor before the sum and
         gradient_predivide_factor / size after the sum.
+      average_aggregated_gradients:
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py:481]: We have seen some use cases where users prefer to manage these reduction via other techniques such as loss scaling but I do agree that if gradients are being averaged likely they want to average both across workers and across aggregations. So if we want to avoid adding an additional argument I am open to just using `op` to set this behavior.

File: horovod/common/gradient_aggregation.py [hunk 4]
```diff
@@ -0,0 +1,266 @@
...
+            sparse_as_dense,
+            average_aggregated_gradients,
+            rank,
+            optimizer_type):
+        self._allreduce_grads = allreduce_func
+
+        # backward_passes_per_step controls how often gradient updates are
+        # synchronized.
+        self.backward_passes_per_step = backward_passes_per_step
+        assert self.backward_passes_per_step > 0
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation.py]: Done.

File: horovod/common/gradient_aggregation.py [hunk 5]
```diff
@@ -0,0 +1,266 @@
...
+            self.counter = tf.compat.v1.get_variable(
+                "aggregation_counter", shape=(), dtype=tf.int32,
+                trainable=False, initializer=tf.compat.v1.zeros_initializer(),
+            )
+            for idx, grad in enumerate(grads):
+                # Handle IndexedSlices.
+                if self.sparse_as_dense and isinstance(grad, tf.IndexedSlices):
+                    grad = tf.convert_to_tensor(grad)
+                elif isinstance(grad, tf.IndexedSlices):
+                    raise AssertionError(
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation.py]: Done

File: horovod/common/gradient_aggregation.py [hunk 6]
```diff
@@ -0,0 +1,266 @@
...
+            clear_op = grad_aggregator.assign(grad_aggregator.initial_value)
+            clear_ops_list.append(clear_op)
+        return tf.group(*clear_ops_list)
+
+    def _aggregate_grads(self, grads):
+        aggregation_ops_list = []
+        grads = get_not_none_from_list(grads)
+        assert len(grads) == len(self.locally_aggregated_grads)
+
+        # Apply new gradient updates to the locally copy.
```
[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/common/gradient_aggregation.py]: Thanks for catching this.

File: horovod/tensorflow/__init__.py (hunk 3 above)
[the remark this answers is not in the document]
Comment (tgaddair) [reply horovod/tensorflow/__init__.py:481]: I think just using `op` is fine to avoid confusion.  If we kept this parameter, we would also need to think about how it applies to Adasum.  For now, if the user wants to do custom reductions, they should be able to set `op=Sum` and then handle it elsewhere.

[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py:481]: If we are switching over to using `op` to toggle averaging aggregated gradients, should I also add `op` as an argument for the [Keras optimizer](https://github.com/horovod/horovod/blob/56183ca42c43aad7afd619f0cc8bc4842336f3ec/horovod/tensorflow/keras/__init__.py#L48) to make it configurable.

[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py:481]: One other factor worth considering here is that if we switch over to using `op` to control averaging across `backewards_passes_per_step` this would be inconsistent with the current behavior of the PyTorch[ _DistributedOptimier](https://github.com/horovod/horovod/blob/56183ca42c43aad7afd619f0cc8bc4842336f3ec/horovod/torch/optimizer.py#L32).

[the remark this answers is not in the document]
Comment (tgaddair) [reply horovod/tensorflow/__init__.py:481]: Good catch.

Regarding Keras: I think it was an oversight that we don't expose `op` as a param to the Keras optimizer, so making that change sounds reasonable to me.

Regarding PyTorch: that's a very a good point.  Currently, this is handled in user code [here](https://github.com/horovod/horovod/blob/master/examples/pytorch/pytorch_imagenet_resnet50.py#L83).

Looking at Adasum, it looks like they don't do any averaging across the local aggregations either.

I think keeping `average_aggregated_gradients` may be the best way to go. However, we may want to set it to `False` by default, and raise an exception for now when the user sets it to `True` along with `op=Adasum`.  Otherwise, the user will get inconsistent behavior between `Average` and `Adasum` when setting `backward_passes_per_step > 1`.  Would that work for your users?

[the remark this answers is not in the document]
Comment (aaron276h) [reply horovod/tensorflow/__init__.py:481]: Yeah this approach makes a lot of sense and allows the most consistency across the different optimizer types. I will go ahead and add `op` as a configuration to the Keras DistributedOptimizer, set the default for `average_aggregated_gradients` to `False` and add a check for  when `op=Adasum` to make sure `average_aggregated_gradients` is set to False. I think down the road it probably makes sense to add `average_aggregated_gradients` to the PyTorch optimizer as well.

[the remark this answers is not in the document]
Comment (tgaddair) [reply horovod/tensorflow/__init__.py:481]: I agree, adding it as a feature for other optimizers / modes would be a great addition in the future.  Thanks for digging into this!

**Comment by aaron276h** [comment]
@tgaddair do you have any suggestions for why I am failing ` Buildkite Unit Tests`? The error message doesn't seem related to the changes in this PR.

**Comment by EnricoMi** [comment]
@aaron276h @tgaddair you are right, the failure has nothing to do with your PR. It is fixed now. Sorry for the confusion.

**Review by tgaddair — APPROVED** [review]
LGTM!  Awesome work.
